stan-dev/stan
# Feature/2759 model base

#### Submission Checklist


#### Summary

Add `stan::model::model_base` base class and and `stan::model::model_crtp` static adapter class for constructing models.   

`model_base` defines pure virtual methods for all functionality used by the services and algorithms.

Model classes are now generated to extend `model_crtp` rather than `prob_grad`.  They also define a factory function to create new instances.  The implementation of the CRTP ensures that all the existing template functions *and* virtual functions are implemented in `model_base` by delegating to the implementation's template function.  (There's an exmaple in the tests for `model_crtp`.)

#### Intended Effect

Speed up compile time.  Specifically, allow functions like the services to be instantiated with  `model_base` and compiled in their own translation units.  Then models are compiled in their own translation unit and linked.


#### How to Verify

Check the matching branch in `stan-dev/cmdstan` that implements CmdStan using the new pattern . You can verify there that there is no slowdown.

Otherwise, there are unit tests for the instantiation and inheritance of the base classes and all other tests still pass.

#### Side Effects

This change maintains backward compatibility in that no functionality is lost from generated model code, which can still be used as before.  All template programs were left intact---they can simply be instantiated with `base_model` rather than being hard coded for `base_model`.  

The main side effect is that all model classes `foo` now contain virtual functions where there used to be non-virtual functions.  As none of these are in tight loops, there is negligible effect on speed.

The `model_name()` function was changed from `static` to a virtual member function so that it could be used as part of the base model specification.  Nothing in the services use the model name, but it might get used in RStan, PyStan, etc.  

#### Documentation

Yes.  The goal is to have the `model_base` fully documented so that clients can understand Stan's requirements for models.

Also added documentation for `prob_grad` that was previously missing.  

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
